src/code/padmgr.c

```diff
@@ -1,58 +1,788 @@
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/padmgr/func_8017544C.s")
+/**
+ * Relinquishes access to the serial message queue, allowing another process to acquire and use it.
+ *
+ * @param serialEventQueue The serial message queue acquired by `PadMgr_AcquireSerialEventQueue`
+ *
+ * @see PadMgr_AcquireSerialEventQueue
+ */
+void PadMgr_ReleaseSerialEventQueue(OSMesgQueue* serialEventQueue) {
+    osSendMesg(&sPadMgrInstance->serialLockQueue, (OSMesg)serialEventQueue, OS_MESG_BLOCK);
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/padmgr/PadMgr_RumbleSet.s")
+/**
+ * Like `PadMgr_ReleaseSerialEventQueue` but for Voice access.
+ *
+ * @see PadMgr_VoiceAcquireSerialEventQueue
+ */
+void PadMgr_VoiceReleaseSerialEventQueue(OSMesgQueue* serialEventQueue) {
+    osSendMesg(&sPadMgrInstance->serialLockQueue, (OSMesg)serialEventQueue, OS_MESG_BLOCK);
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/padmgr/PadMgr_ControllerHasRumblePak.s")
+/**
+ * Locks controller input data while padmgr is reading new inputs or another thread is using the current inputs.
+ * This prevents new inputs overwriting the current inputs while they are in use.
+ *
+ * @see PadMgr_UnlockPadData
+ */
+void PadMgr_LockPadData(void) {
+    osRecvMesg(&sPadMgrInstance->lockQueue, NULL, OS_MESG_BLOCK);
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/padmgr/Padmgr_CalcStickEdges.s")
+/**
+ * Unlocks controller input data, allowing padmgr to read new inputs or another thread to access the most recently
+ * polled inputs.
+ *
+ * @see PadMgr_LockPadData
+ */
+void PadMgr_UnlockPadData(void) {
+    osSendMesg(&sPadMgrInstance->lockQueue, NULL, OS_MESG_BLOCK);
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/padmgr/Padmgr_ParseState.s")
+/**
+ * Activates the rumble pak for all controllers it is enabled on, stops it for all controllers it is disabled on and
+ * attempts to initialize it for a controller if it is not already initialized.
+ */
+void PadMgr_UpdateRumble(void) {
+    s32 i;
+    s32 ret;
+    OSMesgQueue* serialEventQueue = PadMgr_AcquireSerialEventQueue();
+    s32 triedRumbleComm = false;
+    s32 motorStart = MOTOR_START; // required for matching?
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/padmgr/func_801759BC.s")
+    for (i = 0; i < MAXCONTROLLERS; i++) {
+        if (sPadMgrInstance->ctrlrType[i] == PADMGR_CONT_NORMAL) {
+            // Check status for whether a controller pak is connected
+            if (sPadMgrInstance->padStatus[i].status & CONT_CARD_ON) {
+                if (sPadMgrInstance->pakType[i] == CONT_PAK_RUMBLE) {
+                    if (sPadMgrInstance->rumbleEnable[i]) {
+                        if (sPadMgrInstance->rumbleTimer[i] < 3) {
+                            // This should be the osMotorStart macro, however the temporary variable motorStart is
+                            // currently required for matching
+                            if (__osMotorAccess(&sPadMgrInstance->rumblePfs[i], motorStart) != 0) {
```

Comment:
![despair](https://github.com/zeldaret/mm/assets/7416381/e20793c2-0cd6-492e-baaf-b23858b1236b)

Reply:
Based on ultralib, `__osMotorAccess`'s second arg should be `s32` which lets you remove the temp.